examples_tests/51.RadixSort/main.cpp

```diff
@@ -0,0 +1,269 @@
+{
+	ScanClass::Parameters_t scan_push_constants;
+	ScanClass::DispatchInfo_t scan_dispatch_info;
+	ScanClass::buildParameters(in_count, &scan_push_constants, &scan_dispatch_info, WG_SIZE);
+
+	for (uint32_t pass = 0; pass < scan_dispatch_info.upsweep_pass_count; ++pass)
+	{
+		ScanClass::prePassParameterUpdate(pass, true, &scan_push_constants, &scan_dispatch_info, WG_SIZE);
+
+		ScanClass::updateDescriptorSet(ds_upsweep, in_gpu, driver);
+
+		driver->bindComputePipeline(upsweep_pipeline);
+		driver->bindDescriptorSets(video::EPBP_COMPUTE, upsweep_pipeline->getLayout(), 0u, 1u, &ds_upsweep, nullptr);
+		ScanClass::dispatchHelper(upsweep_pipeline->getLayout(), scan_push_constants, scan_dispatch_info, driver);
+
+		ScanClass::postPassParameterUpdate(pass, true, &scan_push_constants, &scan_dispatch_info, WG_SIZE);
+	}
+
+	for (uint32_t pass = 0; pass < scan_dispatch_info.downsweep_pass_count; ++pass)
+	{
+		ScanClass::prePassParameterUpdate(pass, false, &scan_push_constants, &scan_dispatch_info, WG_SIZE);
+
+		ScanClass::updateDescriptorSet(ds_downsweep, in_gpu, driver);
+
+		driver->bindComputePipeline(downsweep_pipeline);
+		driver->bindDescriptorSets(video::EPBP_COMPUTE, downsweep_pipeline->getLayout(), 0u, 1u, &ds_downsweep, nullptr);
+		ScanClass::dispatchHelper(downsweep_pipeline->getLayout(), scan_push_constants, scan_dispatch_info, driver);
+
+		ScanClass::postPassParameterUpdate(pass, false, &scan_push_constants, &scan_dispatch_info, WG_SIZE);
+	}
+}
+
+static void RadixSort(IVideoDriver* driver, const SBufferRange<IGPUBuffer>& in_gpu_range,
+	IGPUDescriptorSet* ds_histogram, IGPUComputePipeline* histogram_pipeline,
+	IGPUDescriptorSet*  ds_upsweep, IGPUComputePipeline*  upsweep_pipeline,
+	IGPUDescriptorSet* ds_downsweep, IGPUComputePipeline* downsweep_pipeline,
+	IGPUDescriptorSet* ds_scatter, IGPUComputePipeline* scatter_pipeline)
+{
+	RadixSortClass::Parameters_t radix_sort_push_constants;
+	RadixSortClass::DispatchInfo_t radix_sort_dispatch_info;
+	RadixSortClass::buildParameters(in_gpu_range.size / sizeof(SortElement), &radix_sort_push_constants, &radix_sort_dispatch_info, WG_SIZE);
+
+	SBufferRange<IGPUBuffer> scratch_gpu_range = { 0 };
+	scratch_gpu_range.size = in_gpu_range.size;
+	scratch_gpu_range.buffer = driver->createDeviceLocalGPUBufferOnDedMem(in_gpu_range.size);
+
+	SBufferRange<IGPUBuffer> histogram_gpu_range = { 0 };
+	histogram_gpu_range.size = radix_sort_dispatch_info.histogram_count * sizeof(uint32_t);
+	histogram_gpu_range.buffer = driver->createDeviceLocalGPUBufferOnDedMem(histogram_gpu_range.size);
+
+	for (uint32_t pass = 0; pass < RadixSortClass::PASS_COUNT; ++pass)
+	{
+		RadixSortClass::updateParameters(&radix_sort_push_constants, pass);
+
+		(pass == 0u)
+			? RadixSortClass::updateDescriptorSet(ds_histogram, { (pass % 2) ? scratch_gpu_range : in_gpu_range, histogram_gpu_range }, driver)
+			: RadixSortClass::updateDescriptorSet(ds_histogram, { (pass % 2) ? scratch_gpu_range : in_gpu_range }, driver);
```

Comment:
you should be able to precompute all these, instead of dynamically updating them

Reply:
also I prefer proper if-statements instead of ternary ops for this